### ISSUE TITLE
unindexObject monkey patch leaves object indexed in some cases

### ISSUE THREAD

**Issue by ewohnlich** [collective/collective.elasticsearch#76]
I am providing steps to reproduce one specific case but the scope is probably broader. The issue occurs for me in both Plone 5.1 and 5.2, using collective.elasticsearch 3.0.3.

**What I did**
1. Create two folders, "Source" and "Destination"
2. In the control panel Content Rules section, create a new rule with triggering event "Workflow state changed"
3. Add condition: Workflow transition is Approve // Publish
4. Add action: move to folder "Destination"
5. Apply this rule to the "Source" folder
6. Create a Page in Source and publish it

**What I expected to happen:**
Catalog should have an entry for /Plone/destination/my-page

**What actually happened:**
Catalog has entries for both /Plone/destination/my-page and /Plone/source/my-page. 

Disabling this monkey patch (unindexObject) resolves the issue. Calling this method by notifying of events that happened is an important part of plone.app.contentrules and it appears to otherwise function correctly. I have no proposed solution because I do not understand the problem that the monkey patch was intended to solve. Is this still needed in Plone 5.1 and 5.2?

**Comment by ewohnlich** [comment]
I think I understand the intention behind the patches - you need the full object when calling the ES hook to remove it from ES (in addition to the ZCatalog). I think this is a problem with versions >= 5.1 because of the change in catalog optimization. The method "uncatalogObject" is now basically an optimization wrapper for "_uncatalogObject" which is most similar to the functionality that this package originally tried to monkey patch.

Indeed if I change the monkey patch ZCML registration from original="uncatalogObject" to original="_uncatalogObject" the problem appears to go away. Instead of trying to make functionality conditional on Plone version, can we instead clean up the monkey patches a little? Right now there are multiple related patches for CatalogTool's unindexObject and uncatalog_object that change the API to use the full obj as a parameter. Basically by changing the API in one you have to patch the other to match the new API, introducing more complexity. Could we instead patch just uncatalog_object and keep its API (and in ElasticSearchCatalog) the same as in CatalogTool, and have ElasticSearchCatalog instead retrieve the object by traversing the UID? It looks like this is already the attempt when dealing with Archetypes.

Something like this (untested)

```
class ElasticSearchCatalog(object):
    ... 

   def uncatalog_object(self, uid):
        # always need to uncatalog to remove brains, etc
        try:
            obj = api.content.get(uid)
        except KeyError:
            pass
        else:
            result = self.catalogtool._old_uncatalog_object(uid)
            if self.enabled:
                hook.remove_object(self, obj)
    
            return result
```

That would use the existing zcml patch for uncatalog_object and the relevant code in patches.py would just need to be tweaked to meet new API. Does this seem like the right idea? I can make an actual PR if it sounds like it's on the right track.

**Comment by idgserpro** [comment]
>Indeed if I change the monkey patch ZCML registration from original="uncatalogObject" to original="_uncatalogObject" the problem appears to go away.

@ewohnlich the path is in [unindexObject](https://github.com/collective/collective.elasticsearch/blob/9f5d42e031997069ae414dd4b12edab93e0ccaaa/src/collective/elasticsearch/configure.zcml#L111). Did you mean unindexObject?

**Comment by ewohnlich** [comment]
Yes, sorry, conflating unindexObject and uncatalog_object

**Comment by idgserpro** [comment]
@ewohnlich I agree to remove the patch from the `unindexObject` method and update the `uncatalog_object` interface. The patch in `unindexObject` seems invasive and unnecessary.

I think the first thing to do is create a PR to fix the build. In PR #78 , we saw that the build is broken.

Then, you can create another PR, in which the first thing is to do a test that demonstrates the error indicated in this issue. You can take a look at the [plone.app.contentrules tests](https://github.com/plone/plone.app.contentrules/blob/master/plone/app/contentrules/tests/test_action_move.py).

PR is welcome!

**Comment by ewohnlich** [comment]
I think I see why it was done the previous way. In attempting to patch only uncatalog_object I can fix the issue here, but ES would no longer set a remove hook correctly. We can't just query the object from path because it's already deleted at this point. The hook need's the object UUID but all we have here is the path to a now deleted object.

Side note for clarification: Plone's CatalogTool refers to "uid" which is really just the Zope path for the object. This allows it to remove catalog entries from content that is already deleted. This is different from the UUID that collective.elasticsearch uses where you really do need to have the full object to get its UUID. Unfortunate.

I think I can still have a fix for my issue and provide a test for it by looking to do the _unindexObject fix I mentioned earlier.

**Comment by ewohnlich** [comment]
@idgserpro I have a proposed fix in https://github.com/ewohnlich/collective.elasticsearch. I don't understand the issue with #78, do you want me to submit a PR anyway?

This fork does not attempt to do any significant refactor as originally intended due to the ES hook's not being able to get the obj UUID as described above. All it does is change the monkey patch of unindexObject to _unindexObject if >= Plone 5.1. I include a changelog entry and a test that requires this fix to pass. Please note I did have to include processQueue commands in two other tests - I don't think those were disrupted by my change, they just need this call to pass in >= Plone 5.1

**Comment by idgserpro** [comment]
@ewohnlich yes, do the PR. But first, it's necessary make a PR just to fix Travis, that is, without any fix code for this issue. Only to make the buildout run.

**Comment by idgserpro** [comment]
@ewohnlich you will need to delete your fork and create a new one to fix the Travis build. One piece of advice I give you is to always create a branch for a fix, even if you are on a fork.

**Comment by idgserpro** [comment]
Fix in #80
